services/typeddata/hash.go

```diff
@@ -0,0 +1,214 @@
+package typeddata
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	bytes32Type, _ = abi.NewType("bytes32")
+)
+
+func deps(target string, types Types) []string {
+	unique := map[string]struct{}{}
+	unique[target] = struct{}{}
+	visited := []string{target}
+	deps := []string{}
+	for len(visited) > 0 {
+		current := visited[0]
+		fields := types[current]
+		for i := range fields {
+			f := fields[i]
+			if _, defined := types[f.Type]; defined {
+				if _, exist := unique[f.Type]; !exist {
+					visited = append(visited, f.Type)
+					unique[f.Type] = struct{}{}
+				}
+			}
+		}
+		visited = visited[1:]
+		deps = append(deps, current)
+	}
+	sort.Slice(deps[1:], func(i, j int) bool {
+		return deps[1:][i] < deps[1:][j]
+	})
+	return deps
+}
+
+func typeString(target string, types Types) string {
+	b := new(bytes.Buffer)
+	for _, dep := range deps(target, types) {
+		b.WriteString(dep)
+		b.WriteString("(")
+		fields := types[dep]
+		first := true
+		for i := range fields {
+			if !first {
+				b.WriteString(",")
+			} else {
+				first = false
+			}
+			f := fields[i]
+			b.WriteString(f.Type)
+			b.WriteString(" ")
+			b.WriteString(f.Name)
+		}
+		b.WriteString(")")
+	}
+	return b.String()
+}
+
+func typeHash(target string, types Types) (rst common.Hash) {
+	return crypto.Keccak256Hash([]byte(typeString(target, types)))
+}
+
+func encodeData(target string, data map[string]interface{}, types Types) (rst common.Hash, err error) {
```

Comment:
I'd suggest to keep the same meaning of `encodeData`, as [in the EIP](url
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-encodedata). Now, it looks like, it's a `hashStruct` (according to EIP).

So it's probably better to rename it to `hashStruct()`. and separate `encodeData` into separate function/method.

Also, as currently used in `hashToSign()` function, there is a hint that it can be also a method of `TypedData`, but I found at least one usage that is not a good fit for that. But perhaps, with the suggested changes, it'll be possible to move from function to method.